### PR TITLE
Use sequences from SPAdes

### PR DESCRIPTION
**Pull request by wafemand**
Sequences from SPAdes use 2 bits per ATGC symbol.

TODOs:
- fix tests
- support N symbol using some sparse data structure
- support other symbols for rare cases
- time profiling

Approximately memory consumption for a particular file. It is based on capacities of containers and `sizeof`s of structs. The input file size is ~80 MB.
|                                           |        old |        new |
|:------------------------------------------|-----------:|-----------:|
| nodes                                     | 250440     | 250440     |
| edges                                     | 131870     | 131870     |
| paths                                     | 107238     | 107238     |
| names heap memory (KB)                           |   7082.62  |   7082.62  |
| sequences heap memory (KB)                         | 134953     |  18325.7   |
| DeBruijnNode structs (KB)                      |  52827.2   |  50870.6   |
| total memory for all nodes (KB)               | 197113     |  78529.1   |
| Total memory for graph (KB)                                    | 242828     | 124244     |

Now the memory bottleneck is a huge `DeBruijnNode` struct (`sizeof(DeBruijnNode) == 216`).

### CODE REVIEW

**Comment by asl** [comment]
Any statistics? Memory consumption? Performance improvements?

File: thirdparty/seq/sequence.hpp [hunk 1]
```diff
@@ -0,0 +1,420 @@
...
+
+        const ST *data() const { return getTrailingObjects<ST>(); }
+        ST *data() { return getTrailingObjects<ST>(); }
+    };
+
+    size_t size_ : 32;
+    size_t from_ : 31;
+    bool   rtl_  : 1;  // Right to left + complimentary (?)
+    llvm::IntrusiveRefCntPtr<ManagedNuclBuffer> data_;
+    std::shared_ptr<llvm::SparseBitVector<>> empty_nucls_ = nullptr;
```
Comment (asl) [inline thirdparty/seq/sequence.hpp]: Ouch. Maybe include this into nuclbuffer then? So we won't have 2 reference counters in the single sequence? :)

File: thirdparty/seq/sequence.hpp [hunk 2]
```diff
@@ -0,0 +1,420 @@
...
+    }
+
+    char getNuclFromBuffer(size_t idx) const {
+        const ST *bytes = data_->data();
+        return static_cast<char>((bytes[idx >> STNBits] >> ((idx & (STN - size_t{1})) << size_t{1})) & size_t{3});
+    }
+
+    bool emptyNuclsEqual(const Sequence &that) const {
+        return (empty_nucls_ == that.empty_nucls_ || (
+                    empty_nucls_ != nullptr
```
Comment (asl) [inline thirdparty/seq/sequence.hpp]: code style

File: graph/assemblygraph.cpp [hunk 3]
```diff
@@ -832,13 +843,11 @@ void AssemblyGraph::makeReverseComplementNodeIfNecessary(DeBruijnNode * node)
...
-        if (node->sequenceIsMissing())
-            nodeSequence = "*";
-        else
+        Sequence nodeSequence{};
+        if (!node->sequenceIsMissing())
             nodeSequence = node->getSequence();
-        DeBruijnNode * newNode = new DeBruijnNode(reverseComplementName, node->getDepth(),
-                                                  getReverseComplement(nodeSequence),
+        auto newNode = new DeBruijnNode(reverseComplementName, node->getDepth(),
+                                                  nodeSequence.GetReverseComplement(),
```
Comment (asl) [inline graph/assemblygraph.cpp]: indent

File: graph/assemblygraph.cpp [hunk 4]
```diff
@@ -1046,13 +1055,13 @@ int AssemblyGraph::buildDeBruijnGraphFromAsqg(QString fullFileName)
                     nodeName = "node";
                 nodeName += "+";
 
-                QByteArray sequence = lineParts.at(2).toLocal8Bit();
-                int length = sequence.length();
+                Sequence sequence{lineParts.at(2).toLocal8Bit()};
+                int length = static_cast<int>(sequence.size());
```
Comment (asl) [inline graph/assemblygraph.cpp:1059]: maybe fix length in node ctor? Why it is signed?

Comment (wafemand) [reply graph/assemblygraph.cpp:1059]: Maybe, the author followed Google Style Guide :)
https://google.github.io/styleguide/cppguide.html#Integer_Types

File: graph/assemblygraph.h [hunk 5]
```diff
@@ -100,7 +101,7 @@ class AssemblyGraph : public QObject
     void buildDeBruijnGraphFromFastg(QString fullFileName);
     void buildDeBruijnGraphFromTrinityFasta(QString fullFileName);
     int buildDeBruijnGraphFromAsqg(QString fullFileName);
-    void buildDeBruijnGraphFromPlainFasta(QString fullFileName);
+    void buildDeBruijnGraphFromPlainFasta(const QString& fullFileName);
```
Comment (asl) [inline graph/assemblygraph.h:104]: Should we fix other methods while here? :)

File: graph/assemblygraphfast.cpp [hunk 6]
```diff
@@ -81,15 +84,16 @@ void AssemblyGraph::buildDeBruijnGraphFromGfa(const QString &fullFileName,
         //that's not available, use a length of 0.
         //If there is a sequence, then the LN tag will be ignored.
         int length;
-        if (sequence == "*" || sequence == "") {
+        if (sequence_bytes == "*" || sequence_bytes == "") {
             auto lnTag = tagWrapper.getNumberTag<int>("LN");
             if (!lnTag) {
-                throw AssemblyGraphError("expected LN tag because sequence is " + sequence);
+                throw AssemblyGraphError("expected LN tag because sequence_bytes is " + sequence_bytes);
```
Comment (asl) [inline graph/assemblygraphfast.cpp]: I would leave `sequence` in the message here ;)

File: thirdparty/seq/sequence.hpp [hunk 7]
```diff
@@ -0,0 +1,420 @@
...
+    template<typename S>
+    void InitEmptyNucls(const S &s, bool rc = false) {
+        int int_size = static_cast<int>(size_);
+        int start = rc ? int_size - 1 : 0;
+        int end = rc ? -1 : int_size;
+        int step = rc ? -1 : 1;
+        int cur_index = 0;
+
+        for (int i = start; i != end; i += step) {
+            if (is_N(s[i])) {
```
Comment (asl) [inline thirdparty/seq/sequence.hpp]: I would put here `unlikely`

**Review by asl — APPROVED** [review]
LGTM, thanks!

**Pull request merged**